lib/elixir/lib/module/checker.ex

```diff
@@ -3,58 +3,93 @@ defmodule Module.Checker do
 
   @moduledoc false
 
-  def verify(module_map, binary, cache) do
-    warnings = warnings(module_map, cache)
-    binary = chunk(binary, module_map)
-    {binary, warnings}
+  def verify(module, cache) do
+    case prepare_module(module) do
+      {:ok, map} -> {:ok, build_chunk(map), warnings(map, cache)}
+      :error -> :error
+    end
+  end
+
+  defp prepare_module({_module, map}) when is_map(map) do
+    {:ok, map}
+  end
+
+  defp prepare_module({module, binary}) when is_binary(binary) do
+    with {:ok, debug_info} <- debug_info(module, binary),
+         {:ok, checker_info} <- checker_chunk(binary) do
```

Comment:
I am not a fan of the `Map.merge` below. I generally like to treat those maps as "static"  constructs: i.e. all of the keys will be define only once and all upfront. So I would make the functions above return things like `{:ok, definitions, ...}` and then create the map upfront.